### PR TITLE
RoE: Miscellaneous Improvements + Fixes

### PR DESCRIPTION
**Pull request by Kreidos**
Just some simple fixes.

-Adds missing comments and improves commenting
-Adds one record which was marked TODO
-Performance Optimizes GetNumEminenceCompleted & corrects it to not count hidden record IDs
-Removes old workaround for All For One
-Fixes blue caskets not counting towards Crack Caskets record
-Timed vanquish records now exclude Abyssea areas 

**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

### CODE REVIEW

File: scripts/globals/roe.lua [hunk 1]
```diff
@@ -243,10 +249,7 @@ local function completeRecord(player, record)
...
     then
         local bonusAccoladeRate = 1.0
-
-        if record ~= 5 then -- Do not grant a bonus for All for One
-            bonusAccoladeRate = bonusAccoladeRate + ((player:getUnityRank() - 1) * 0.05)
-        end
+        bonusAccoladeRate = bonusAccoladeRate + (player:getUnityRank() or 0) * 0.05
 
         player:addCurrency("unity_accolades", math.floor(rewards["accolades"] * bonusAccoladeRate), CAP_CURRENCY_ACCOLADES)
         player:messageBasic(xi.msg.basic.ROE_RECEIVED_ACCOLADES, rewards["accolades"], player:getCurrency("unity_accolades"))
```
Comment (Kreidos) [inline scripts/globals/roe.lua]: Should this Accolades received message be showing the actual amount you received after the bonus calculation? Or the pre-bonus amount as it does now? :thinking:

Comment (claywar) [reply scripts/globals/roe.lua]: I'll agree that it should be showing full; however, Record ID 5 will not receive this bonus (static 1000 value).  Also, Rank 1 unity receives no bonus, and that logic is removed in this PR.

Comment (Kreidos) [reply scripts/globals/roe.lua]: Actually, I checked that. It appears as though the "subtract 1 from unity rank" is already being done on the `getUnityRank()` end. So it was in fact subtracting 5% for being Unity rank 1 before. This was likely never caught because the message always showed the flat pre-bonus reward.

Comment (claywar) [reply scripts/globals/roe.lua]: Its a little confusing, but unity leader is 1-indexed with "None" being 0.  The rank values in the DB are also reflective of their actual ranking (not 0-index)

Comment (Kreidos) [reply scripts/globals/roe.lua]: I see. Reinstated the -1.

File: scripts/globals/roe.lua [hunk 2]
```diff
@@ -166,6 +166,8 @@ local defaults = {
                                 --        "weekly" - Weekly record.
                                 --        "unity"  - Weekly reset, but doesn't reset progress, only completion.
                                 --        "retro"  - Can be claimed retroactively. Calls check on taking record.
+                                --        "hidden" - Special internal-use record used only as a client-flag to unlock others.
```
Comment (claywar) [inline scripts/globals/roe.lua:169]: Hidden Records are handled client-side with no impact to core or the server.  Is there a reason for separating this into an additional bitfield?

The records absolutely exist, but they are not displayed in the client.

Comment (Kreidos) [reply scripts/globals/roe.lua:169]: Indeed, you are correct on all counts. However, those records shouldn't be counted by The function which counts how many records a player completed.

File: scripts/globals/roe.lua [hunk 3]
```diff
@@ -249,7 +249,7 @@ local function completeRecord(player, record)
         type(rewards["accolades"]) == "number"
     then
         local bonusAccoladeRate = 1.0
-        bonusAccoladeRate = bonusAccoladeRate + (player:getUnityRank() or 0) * 0.05
+        bonusAccoladeRate = bonusAccoladeRate + ((player:getUnityRank() or 1) - 1) * 0.05
```
Comment (claywar) [inline scripts/globals/roe.lua]: This still doesn't resolve RoE 5, since the unity is set before this code is hit.  `CLuaBaseEntity::getUnityRank()` will only return null if no leader has been chosen, but this call happens after the first leader is set (onEventFinish in unity.lua global).

Comment (claywar) [reply scripts/globals/roe.lua]: Other than that, everything else looks good to me 👍

Comment (claywar) [reply scripts/globals/roe.lua]: Easy fix, move trigger:
```lua
    -- First time joining Unity (Requirements met for num Objectives and All for One set)
    if csid == zoneEventIds[zoneId][3] and option >= 1 and option <= 11 then
        changeUnityLeader(player, option)
        xi.roe.onRecordTrigger(player, 5)
        player:messageSpecial(ID.text.YOU_HAVE_JOINED_UNITY, option - 1)
```
Trigger before changing leader, and the previous logic can stand.

Comment (Kreidos) [reply scripts/globals/roe.lua]: Done. I approve of dealing with edge-cases within their respective areas. :+1:

**Review by claywar — APPROVED** [review]

**Comment by claywar** [comment]
Just realized that by moving the accolade reward above setting a unity leader that while it will fix the issue of bonus being applied, it won't track against the player's new unity ranking.

**Comment by Kreidos** [comment]
Has that behaviour been confirmed as accurate? It seems pretty odd that the bonus doesn't apply but somehow those points still count towards the unity ranking.

**Comment by claywar** [comment]
So when you complete all for one, a player receives a flat 1000 accolades and it does apply to unity ranking. That was the previous logic, and the reason for the conditional. This is the only time that behavior occurs

File: src/map/roe.cpp [hunk 4]
```diff
@@ -266,23 +273,25 @@ namespace roeutils
...
+            for ( ; pageVal; completedCount += !RoeSystem.HiddenRecords.test(page * 8 + bitIndex))
             {
-                completedCount += pageVal & 1;
-                pageVal >>= 1;
+                #ifdef _MSC_VER
+                    _BitScanForward(&bitIndex, pageVal);
+                #else
+                    bitIndex = __builtin_ctz(pageVal);
+                #endif
+                pageVal &= (pageVal - 1);
```
Comment (zach2good) [inline src/map/roe.cpp]: If the logic here is:
```
while pageVal is non-zero:
    scan across it and test each set bit
    accumulate result into completedCount
    then progress to the next page
```
Can we do that with a while loop instead of this hard to parse for loop?

If that simplifies how this looks, it may or may not be useful to extract these intrinsic out into a helper in `common`:
```cpp
// Scans from LSB to MSB
unsigned long ScanForFirstSetBit(uint8 value)
{
    unsigned long bitIndex = 0;
    #ifdef _MSC_VER
        _BitScanForward(&bitIndex, value);
    #else
        bitIndex = __builtin_ctz(value);
    #endif
    return bitIndex;
}
```

Comment (Kreidos) [reply src/map/roe.cpp]: Whatever makes the code easier for others to maintain and understand.

Intrinsics could be broken out if someone wishes to, but I feel their use cases might be relatively niche.

Comment (claywar) [reply src/map/roe.cpp]: To follow up on this, I would argue that we should absolutely keep things like this in an isolated place with pre-defined wrappers.  This is not a question of niche cases, but more an issue of maintainability.  Its much easier to handle environment differences in a single location, than potentially peppering the code with one-off cases.

EDIT: While this isn't the case now, this is a pattern that should be avoided.

File: src/map/roe.cpp [hunk 5]
```diff
@@ -266,23 +273,26 @@ namespace roeutils
         return PChar->m_eminenceLog.complete[page] & (1 << bit);
     }
 
-    uint16 GetNumEminenceCompleted(CCharEntity* PChar)
-    {
-        uint16 completedCount = 0;
+    uint16 GetNumEminenceCompleted(CCharEntity* PChar){
+        uint16 completedCount {0};
```
Comment (zach2good) [inline src/map/roe.cpp]: TODO: Style

**Review by zach2good — APPROVED** [review]
LGTM with one style ding 👍

**Review by zach2good — APPROVED** [review]
